src/core/include/mp-units/bits/sudo_cast.h

```diff
@@ -64,34 +103,77 @@ template<Quantity To, typename From>
-    constexpr Magnitude auto num = numerator(c_mag);
-    constexpr Magnitude auto den = denominator(c_mag);
-    constexpr Magnitude auto irr = c_mag * (den / num);
-    using c_rep_type = maybe_common_type<typename std::remove_reference_t<From>::rep, typename To::rep>;
-    using c_mag_type = common_magnitude_type<c_mag>;
-    using multiplier_type = conditional<
-      treat_as_floating_point<c_rep_type>,
-      // ensure that the multiplier is also floating-point
-      conditional<std::is_arithmetic_v<value_type_t<c_rep_type>>,
-                  // reuse user's type if possible
-                  std::common_type_t<c_mag_type, value_type_t<c_rep_type>>, std::common_type_t<c_mag_type, double>>,
-      c_mag_type>;
-    using c_type = maybe_common_type<c_rep_type, multiplier_type>;
-    constexpr auto val = [](Magnitude auto m) { return get_value<multiplier_type>(m); };
-    if constexpr (std::is_floating_point_v<multiplier_type>) {
+    using traits = magnitude_conversion_traits<To, std::remove_reference_t<From>>;
+    if constexpr (std::is_floating_point_v<typename traits::multiplier_type>) {
       // this results in great assembly
-      constexpr auto ratio = val(num) / val(den) * val(irr);
       auto res = static_cast<MP_UNITS_TYPENAME To::rep>(
-        static_cast<c_type>(q.numerical_value_is_an_implementation_detail_) * ratio);
+        static_cast<traits::c_type>(q.numerical_value_is_an_implementation_detail_) * traits::ratio);
       return {res, To::reference};
     } else {
       // this is slower but allows conversions like 2000 m -> 2 km without loosing data
       auto res = static_cast<MP_UNITS_TYPENAME To::rep>(
-        static_cast<c_type>(q.numerical_value_is_an_implementation_detail_) * val(num) / val(den) * val(irr));
+        static_cast<traits::c_type>(q.numerical_value_is_an_implementation_detail_) * traits::num_mult /
+        traits::den_mult * traits::irr_mult);
       return {res, To::reference};
     }
   }
 }
 
+
+/**
+ * @brief Explicit cast between different quantity_point types
+ *
+ * @note This is a low-level facility and is too powerful to be used by the users directly. They should either use
+ * `value_cast` or `quantity_cast`.
+ *
+ * @tparam ToQP a target quantity point type to which to cast to
+ */
+template<QuantityPoint ToQP, typename FromQP>
+  requires QuantityPoint<std::remove_cvref_t<FromQP>> &&
+           (castable(std::remove_reference_t<FromQP>::quantity_spec, ToQP::quantity_spec)) &&
+           (detail::same_absolute_point_origins(ToQP::point_origin, std::remove_reference_t<FromQP>::point_origin)) &&
+           ((std::remove_reference_t<FromQP>::unit == ToQP::unit &&
+             std::constructible_from<typename ToQP::rep, typename std::remove_reference_t<FromQP>::rep>) ||
+            (std::remove_reference_t<FromQP>::unit != ToQP::unit))
+[[nodiscard]] constexpr QuantityPoint auto sudo_cast(FromQP&& qp)
+{
+  using qp_type = std::remove_reference_t<FromQP>;
```

Comment:
I think you are assuming here an integral representation type, which often will not be the case as `double` is the default in the library. In cases of a floating point representation type, we should calculate the entire conversion factor in one step and then do only one multiplication operation on the value to get the result. Otherwise, we will not generate assembly equivalent to operations on `double`.

Reply:
Indeed, those considerations about order of the conversion operations are most relevant for integral types, where both underflow and overflow are likely. It doesn't hurt for double either. However, I'm not sure what you mean by "calculate the entire conversion factor in one step". A change of point-origin requires an addition. A change in unit requires a multiplication. This implementation here does at most one multiplication and at most one addition in all cases. Here is how:
 - The initial check for same point-origin directly delegates to the `sudo_cast<Q>` implementation to select the "best" way to do that multiplication, and avoids the addition
 - The two code-paths for a change in point origin go as follos:
   - The first path for where the unit gets smaller and thus the number gets larger:
     1. *cast to the intermediate representation type* using `sudo_cast` (no-op in the case of floatingpoints). previously, that was a `value_cast`, but I figured that within a `sudo_cast`, `value_cast` is off limit.
     2. *add/subtract the change in reference* using `.point_for`; hopefully, the implementation of `.point_for` is such that this is a single addition and no multiplication
     3. *scale the unit/number* and 
     4. *cast the to the final representation type* using `sudo_cast`.
   - The second path for where the unit gets larger and thus the number will get smaller (or stay the same):
     1. *cast to the intermediate representation type* and
     2. *scale the unit/number* using `sudo_cast`. If the intermediate representation type is the same as the source erpresentation and the units remain the same, this is a `no_op`.
     3. *add/subtract the change in reference* using `.point_for`
     4. *cast to the final representation type* using `sudo_cast`. No-op for floating-point types.

So I believe this implementation is as efficient as it can be through a careful selection of types. While both code-paths invoke `sudo_cast` twice, there is always one where the source and target units match and no multiplication is needed. For floating-point, there is also not cast, so that cast is a no-op. Furthermore, the implementation here does never do any arithmetic itself - it delegates the addition to `.point_for` and the multiplication to `sudo_cast<Q>`. Personally, I would even prefer to call `value_cast` instead, because in each of the two code-paths, one of the casts is a pure representation cast, which would be more clear as `value_cast<rep>`.

Reply:
Wait, now I understand, you are indicating that `.point_for` may potentially change the unit and representation type? In that case, you are of course right, we should replace the `.point_for` call here with a dedicated implementation.

Reply:
> However, I'm not sure what you mean by "calculate the entire conversion factor in one step".

I mean something like this:

```cpp
  if constexpr (std::is_floating_point_v<multiplier_type>) {
      // this results in great assembly
      constexpr auto ratio = val(num) / val(den) * val(irr);
     // use precalculated ratio...
  }
  else // ...
```

Precalculation of `ratio` as a `constexpr` variable ensures that the assembly will include only one multiplication by this constant. The `else` branch typically results in more assembly instructions.

Reply:
@burnpanck, are you going to work on this here? Or should we rebase, merge, and then refactor in the next PR?

Reply:
Sorry, work interfered. Yeah, I intend to do a direct implementation of `point_for` which avoids all further unit conversions. However, I believe the code as-is would be mergeable too. As mentioned before, there should definitely be no regressions in this PR as of now. All unit conversions are as efficient as they can be (they re-use the existing implementation). The current `value_cast<QP>` is no worse than what you would reasonably implement using the tools given by this library as a user (namely, using `point_for`). The converting constructor of `quantity_point` is untouched and slightly worse than this `value_cast<QP>` concerning overflow or truncation with integer types. Maybe it would make sense to switch the converting constructor to delegate to `value_cast<QP>` and then merge. We can then postpone performance-improvement of `value_cast<QP>` to another PR, and perhaps discuss separately when the current converting constructor of `quantity_cast` should be `explicit` or not.